state/protocol/protocol_state/mutator.go

```diff
@@ -5,57 +5,247 @@ import (
+	m.pendingDbUpdates = append(m.pendingDbUpdates, dbUpdates...)
+	return nil
 }
 
-// CommitProtocolState commits the protocol state updater as part of DB transaction.
-// Has to be called for each block to correctly index the protocol state.
-// No errors are expected during normal operations.
-func (m *Mutator) CommitProtocolState(updater protocol.StateUpdater) func(tx *transaction.Tx) error {
-	return func(tx *transaction.Tx) error {
-		updatedState, updatedStateID, hasChanges := updater.Build()
-		if hasChanges {
-			err := m.protocolStateDB.StoreTx(updatedStateID, updatedState)(tx)
-			if err != nil && !errors.Is(err, storage.ErrAlreadyExists) {
-				return fmt.Errorf("could not store protocol state (%v): %w", updatedStateID, err)
+// handleServiceEvents handles applying state changes which occur as a result
+// of service events being included in a block payload:
+//   - inserting incorporated service events
+//   - updating protocol state for the candidate block
+//
+// Consider a chain where a service event is emitted during execution of block A.
+// Block B contains a receipt for A. Block C contains a seal for block A.
+//
+// A <- .. <- B(RA) <- .. <- C(SA)
+//
+// Service events are included within execution results, which are stored
+// opaquely as part of the block payload in block B. We only validate and insert
+// the typed service event to storage once we process C, the block containing the
+// seal for block A. This is because we rely on the sealing subsystem to validate
+// correctness of the service event before processing it.
+// Consequently, any change to the protocol state introduced by a service event
+// emitted during execution of block A would only become visible when querying
+// C or its descendants.
+//
+// This method will only apply service-event-induced state changes when the
+// input block has the form of block C (ie. contains a seal for a block in
+// which a service event was emitted).
+//
+// All updates that are incorporated in service events are applied to the protocol state by mutating protocol state updater.
+// This method doesn't modify any data from self, all protocol state changes are applied to state updater.
+// All other changes are returned as slice of deferred DB updates.
+//
+// Return values:
+//   - dbUpdates - If the service events are valid, or there are no service events,
+//     this method returns a slice of Badger operations to apply while storing the block.
+//     This includes operations to insert service events for blocks that include them.
+//
+// No errors are expected during normal operation.
+func (m *stateMutator) handleServiceEvents(seals []*flow.Seal) (dbUpdates []func(*transaction.Tx) error, err error) {
+	epochFallbackTriggered, err := m.params.EpochFallbackTriggered()
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve epoch fallback status: %w", err)
+	}
+
+	parentProtocolState := m.stateMachine.ParentState()
+	epochStatus := parentProtocolState.EpochStatus()
```

Comment:
I would expect that we use the `ProtocolStateEntry`'s flag here: https://github.com/onflow/flow-go/blob/f1965e1861dcc2f2e8b23bb60ab36abb758a2635/model/flow/protocol_state.go#L35

Reply:
https://github.com/onflow/flow-go/pull/4868/commits/f57cb5e75f0025417a075de1e8626dbb0f8b7ed9